### ISSUE TITLE
Server is not considering application dependency

### ISSUE THREAD

**Issue by kristian** [SAP/ui5-project#61]
## Expected Behavior
If you are developing on a library (or future component types see related issue https://github.com/SAP/ui5-tooling/issues/45) and you add an application project as a 
`dependency` or `devDependency`, the application dependency should be served as the server root.

Also the dependent applications index.html etc. should be served.

## Current Behavior
Currently the application dependency is not served at all when beeing added as a dependency.

The application appears in ui5 tree, but it'll still not be served.

## Steps to reproduce the issue
1. Create a library project
2. Add a application project as dependency
3. Try to serve

## Affected components
- [ ] [ui5-builder](https://github.com/SAP/ui5-builder)
- [x] [ui5-server](https://github.com/SAP/ui5-server)
- [ ] [ui5-cli](https://github.com/SAP/ui5-cli)
- [ ] [ui5-fs](https://github.com/SAP/ui5-fs)
- [x] [ui5-project](https://github.com/SAP/ui5-project)
- [ ] [ui5-logger](https://github.com/SAP/ui5-logger)

**Comment by RandomByte** [comment]
This is more related to the ui5-project module. The server doesn't necessarily need to care about project types

**Comment by kristian** [comment]
> This is more related to the ui5-project module. The server doesn't necessarily need to care about project types

Hello Merlin, mhm I am not sure about that statement... Just looking at it from an high level, I would have even put it the other way around:

Why should the ui5-project module care about the project type? As I see it, the ui5-project module should build the dependency tree just by following the specific rules of the translator in use. E.g. for the npm-translator, the dependency tree should be built solely by looking at the NPM (dev-)dependencies. 

The project type should not make a difference here... The ui5-project module might just output "this is an application" and "this is a library" in addition when building the tree, but essentially I would not see any additional logic required here.

Furthermore (and again IMHO ["in my humble opinion"], so please just feel free see it diffrently =) ) I would expect the root element of the tree output by the ui5-project module, should always be the project, in which the tree is generated for, again, regardless of the type of the project! So taking the example of two projects, one is a devDependency and the other one is a dependency to one another:

```
projectA has a dependency to projectB
projectB has a devDependency to projectA
```

projectA beeing an application project, projectB beeing a library project. I would expect the tree to look like this:

```
projectA (application)
\- projectB (library)
```

if executed on projectA and

```
projectB (library)
\- projectA (application)
```

if executed on projectB. The ui5-server module could then apply some logic based on the project type (because why should the ui5-project module "assume" that only application projects can be executed by the server?). The logic could be something like:

1. Calcuate the ui5-project depenency tree based on the current module
2. If the root of the tree is an application project serve the application project and all other projects as dependencies
3. In case the root is not an application project, check the dependency tree for application projects:
    * In case there is no application project in the tree: fail with an error
    * In case there is exactly one application project in the tree: re-calculate the dependency tree based on this application project and serve it
    * In case there are multiple application projects in the tree: either fail with an error, or expect e.g. an `--application` parameter for the server, which tells the server which of the ambiguous projects to execute

For me that way arround it makes more sense, as the ui5-server module would imply the logic what can be executed and what not... Also for future use cases it would open up more possibilities in my opinion. So e.g. in reference to https://github.com/SAP/ui5-tooling/issues/45 when introducing a new project type "component", the server might be ehanced, to serve "component" projects directly, by serving a dummy-HTML containing just a ComponentContainer (e.g. via the new `data-sap-ui-oninit="module:sap/ui/core/ComponentSupport"`). Not putting the logic for the component types into ui5-project would, in my opinion, allow such enhancements to be made more easily.

**Comment by kristian** [comment]
Thinking about it, just as one further improvement for my second case. Instead of re-calculating the whole tree a better implementation would be to simply perform some [tree rotation](https://en.m.wikipedia.org/wiki/Tree_rotation) operations, to make the application project the root of the tree. And as this is not a binary tree, you wouldn‘t even have to care about reassigning tree nodes, so the rotation operation would be super quick and by a magnitude faster than re-calculating the whole tree. =)

**Comment by RandomByte** [comment]
Hi Kristian,

Thank you for your input.

I'll try to clarify some points:

> Why should the ui5-project module care about the project type? As I see it, the ui5-project module should build the dependency tree just by following the specific rules of the translator in use.

The **translators** create a mostly unopinionated dependency tree.

The **projectPreprocessor** takes in that dependency tree and enriches it with project specific configurations. These configurations are partially type-dependent. For example a `webapp`-directory configuration is only possible in projects of type `application`. Therefore, type-specific "formatters" are being used to configure the projects.

Finally, the [**normalizer**](https://github.com/SAP/ui5-project#normalizer) uses a translator and the projectPreprocessor to provide a configured dependency tree that can be used by the ui5-builder and ui5-server.

> I would expect the root element of the tree output by the ui5-project module, should always be the project, in which the tree is generated for, again, regardless of the type of the project!

This is the case and will continue to be the case. Technically there is no need for an application project to be the root project. It was just a restriction we put in place for the time being.

**Background:**
application projects do not have a namespace. This means their resources are always mapped to the root (`/`) path. Multiple application projects in the same tree would therefore probably end up having resource clashes (e.g. for two application projects, both `index.html` files would be mapped to the path `/index.html`).

To prevent this we currently ignore all application projects that are not the root project. We will change this with #62 to support use cases like yours. The new logic basically ignores all application projects but the one *closest to the root* (not necessarily the root project).

Note that the mapping between physical and virtual paths is done by **ui5-fs**. ui5-builder and ui5-server facilitate ui5-fs. They throw in a project tree and get a "virtual" file system in return. This means ui5-server does not deal with any path mapping, it just serves the ui5 file system. Currently, ui5-server does not even look at the type of any projects.

> if executed on projectB. The ui5-server module could then apply some logic based on the project type (because why should the ui5-project module "assume" that only application projects can be executed by the server?). The logic could be something like:

ui5-fs doesn't really care about the position of a project in the tree. An application project is always mapped to the root (`/`) (as configured by the ApplicationFormatter during project preprocessing in ui5-project).

Therefore, after the "application project must be root"-restriction has been lifted with #62, starting the server in projectA will have the same effect as starting it in projectB.

**Comment by RandomByte** [comment]
We'll do a release tomorrow

**Comment by RandomByte** [comment]
Fix can be tested with UI5 CLI v0.2.4
